### ISSUE TITLE
Basic features for Yoast 14/15

### ISSUE THREAD

**Issue by herrvigg** [qtranslate/qtranslate-xt#947]
Since i recreated the module from scratch for Yoast 14 (see #794) i suggest we discuss here at least the basic features we can deliver. Currently there is only a template for `front` and `admin`. I didn't bring any javascript code yet since i believe it has change drastically and needs to be refactored completely.

This can be of interest for @pjrobertson who started working on this. This is a bit radical but the goal is to start from a clean point. The old code is available in `qtranslate-xt/dev/wp-seo-yoast-13` as a reference. We may copy relevant code if it is proven that it works.

**Comment by herrvigg** [comment]
To start with, I added `wpseo_opengraph_title` as suggested in #944. This can be done with a PHP filter or via a `i18n-config.json` file. I tend to prefer PHP code but at the same time, the `i18n-config.json` is still the standard way for integration with other plugins so perhaps we should use that as much as possible and then improve the whole framework.

Anyway, there are surely plenty of filters we should define to support Yoast properly. Some were mentioned in #794 but things may have already moved since then. I found this page which seems to list most of the relevant ones: https://developer.yoast.com/customization/apis/metadata-api. We can start to look at this.

On a more complex side, the integration with Javascript code may perhaps start from this: https://developer.yoast.com/customization/yoast-seo/adding-custom-data-analysis. But i don't know yet. 

The main question is how to support the translatable fields. Those that can be handled in a simple fashion can be discussed here. For more complex integration i suggest we create separate tickets so we can follow all the different topics more easily.

**Comment by BenceSzalai** [comment]
I am using Yoast v15 with this plugin:
https://github.com/qTranslate-Team/wp-seo-qtranslate-x/

I don't think it does much useful things as it is very old, apart from containing a fair [`i18n-config.json`](https://github.com/qTranslate-Team/wp-seo-qtranslate-x/blob/master/i18n-config.json) with regards of the `front-config` section, with which as far as I can see the Yoast fields are translated on the frontend. 

The issues are in the Admin UI. But Most of those cannot be fixed with using `i18n-config.json` because Yoast does some JS magic with the input fields.

I've added myself this, so at least the company name works:
```json
...
  "admin-config": {
    "yoast_wpseo_config": {
      "pages": {"admin.php":"page=wpseo_titles"},
      "forms":{
        "wpseo-conf":{
          "fields":{
            "company_name":{"encode":"["}
          }
        }
      }
    },
...
```
![company name](https://i.imgur.com/WIlRPtx.png)


The rest right now can be edited in raw format, e.g.:
![seo fields](https://i.imgur.com/wPrOcdj.png)

**Comment by herrvigg** [comment]
Just a quick remark, in the i18-config.json the values `{"encode":"["}` can be rewritten as `{}` as the `[]` brackets are the default encoding.
For example: `"company_name": {}`

**Comment by herrvigg** [comment]
Also here for the track record: i realized the `i18n-config.json` is not supported natively in the modules. See #958. So i will rather convert this to PHP code in the modules for now.

**Comment by herrvigg** [comment]
Mmm i just realized the previous version of the module was indeed completely broken. The front filters were added on action `i18n_front_config`. But that filter can't be used in the modules... because it's fired _before_ they are loaded (!). This worked only at the time when there was a separate plugin, the call order was different.

It seems Yoast was the only one in that case, but this framework needs to be reworked for a better initialization sequence. For now i just add them directly when the module is loaded.

**Comment by herrvigg** [comment]
As a baseline we have now in the front filters:
- wpseo_metadesc
- wpseo_title
- wpseo_twitter_description
- wpseo_twitter_title
- wpseo_opengraph_desc
- wpseo_opengraph_title

I took this from: https://developer.yoast.com/customization/apis/metadata-api/ and checked what we had in the old version of the module and comments in #794.

Despite the fact that the title cannot be handled yet on the admin side, i expect this to work for the frontside, as we just want to translate the value to be displayed.  So that should be already a good start!

Just one reserve: i noticed that for WooCommerce the front filters are **not** set when a **cron request** is running (`wp_doing_cron()`). The reason given in a comment is:
> Dealing with webhooks, which should always send information in Raw ML format

Perhaps this should also be done for Yoast and other modules, but not sure. If someone can check this it would be easy to fix.

**Comment by herrvigg** [comment]
@BenceSzalai just a few comments
> I am using Yoast v15 with this plugin: https://github.com/qTranslate-Team/wp-seo-qtranslate-x/

This plugin is now much obsolete and it's not compatible with the modules. Could you try with the last version of the module now updated in master? I'm aware many parts may be missing on the admin side, but since Yoast 14 almost everything was broken so we had to restart from scratch (as discussed in #794).

If we can add easily the support for other admin fields that would be really nice.

> The issues are in the Admin UI. But Most of those cannot be fixed with using i18n-config.json because Yoast does some JS magic with the input fields.

Yes, that's a general problem and one of the main limitations we have in qTranslate.

As mentioned for WooCommerce, we need a better framework for the javascript side but i expect this to be much better with a rework using Babel and a more modern JS framework: #887. Let's deal with that separately, unless someone finds a quick solution from the previous plugin/module we can integrate easily right now.

**Comment by pjrobertson** [comment]
Great to see you working on this @herrvigg ! This module is really needed.

I still don't have the time to help with a refactoring, but I'll share what I've done here.

1. Set up a cron job to run this every day. This fixes the fact that Yoast saves only the current language for the 'breadcrumbs' indexable:
```sql
UPDATE wp_yoast_indexable AS yoast
  LEFT JOIN wp_posts AS posts ON yoast.object_id = posts.id
  LEFT JOIN wp_postmeta AS meta ON yoast.object_id = meta.post_id and meta_key = '_yoast_wpseo_bctitle'
  SET yoast.breadcrumb_title = IF(meta.meta_value IS NOT NULL, meta.meta_value, posts.post_title) WHERE yoast.object_type = 'post'
```
2. Added the following to qpseo-front.php to fix the JSON 'schema' output:
```php
add_filter('wpseo_schema_webpage', 'fix_webpage_schema',  10, 2);
function fix_webpage_schema($piece, $context) {
    if (array_key_exists('description', $piece)) {
      $piece['description'] = qtranxf_useCurrentLanguageIfNotFoundUseDefaultLanguage($piece['description']);
    }
  return $piece;
}

add_filter('wpseo_breadcrumb_single_link_info', 'fix_bc_link', 10, 3);
function fix_bc_link($link_info, $index, $crumbs) {
  $link_info['text'] = qtranxf_useCurrentLanguageIfNotFoundUseDefaultLanguage($link_info['text']);
  return $link_info;
}
```


Both of these are hacks, but as I said I don't have the time to dive into this properly at the moment.

**Comment by jurijc** [comment]
@pjrobertson 
Hi I have been currently doing same hacks for other translation plugin (that saves to meta and suits me more) and had same problem, so wanted to share some solution.

If currently is not main language use: 
add_filter('wpseo_should_save_indexable','__return_false');

That prevents saving indexable

**Comment by herrvigg** [comment]
@pjrobertson great you have something there. The second part may be merged but the first part is more problematic. Manipulating the cron and SQL at such low level is a bit scary to generalize and we don't have any cron features in qTranslate so far. It also goes beyond the scope of this initial ticket, so I suggest to open a new one for this.

**Comment by herrvigg** [comment]
@pjrobertson would it make sense to merge the second part without the first one with the cron? If that brings at least some translation feature we could do it.

**Comment by pjrobertson** [comment]
I wouldn’t merge the cron/Mysql part, that’s just a hack to get this working temporarily. The 2nd part can be merged, however.

**Comment by herrvigg** [comment]
@pjrobertson first part merged with the front filters.

I think we can close this ticket for the initial support. I plan to release a new version to github soon with all these recent changes for Yoast. Let me know if something else is missing.

**Comment by herrvigg** [comment]
First batch of changes released in 3.9.3.
I prefer to close this issue, better to create new issues to handle all topics separately.
For the dynamic updates in Javascript this requires more work.
